nil/core-concepts/shards-parallel-execution.mdx

```diff
@@ -29,6 +29,92 @@ As the number of execution shards grows, this process may be changed so that sha
+      'lineColor': '#F8B229',
+      'secondaryColor': '#181A1B',
+      'tertiaryColor': '#fff'
+    }
+  }
+}%%
+sequenceDiagram
+    participant S1 as Shard 1 with Contract 1
+    participant S2 as Shard 2 with Contract 2
+    S1 ->> S1: Contract 1 calls a function<br>inside Contract 2
+    S1 ->> S1: Shard 1 creates a new transaction<br>whose destination is Shard 2
+    Note right of S1: Contract 1 is freed and can execute other operations
+    S1 --> S2: Shard 2 picks up the transaction<br>and adds it to the mempool
+    S2 ->> S2: Shard 2 executes the transaction<br>and includes it in a block
+    S2 ->> S2: Contract 2 executes the function<br>that Contract 1 called   
+```
+
+The key stages of this flow are:
+
+1. Contract 1 calls a function inside Contract 2, which creates a new transaction
+2. Shard 2 must pick up this transaction and place it inside its mempool
+3. When Shard 2 processes the transaction, Contract 2 executes the function that Contract 1 called
+
+:::info
+
+Transactions that originate as a result of cross-shard calls are known as [**cross-shard transactions**](../guides/architecture/ethereum-nil-diffs#transactions). They constitute a separate transaction type in addition to Ethereum-style transactions supported by =nil;.
+
+:::
+
+For an in-depth explanation of how to perform and manage async calls, [**click here**](../getting-started/essentials/handling-async-execution).
+
+:::info[Callback pattern]
+
+There is also a variation of the above flow in which the function inside Contract 2 makes an async call back to Contract 1.
+
+In this case, the flow would work as follows:
+
+1. Contract 1 calls a function inside Contract 2, creating a new transaction
+2. Shard 2 executes the transaction, and Contract 2 calls the requested function
+3. As part of executing the function, Contract 2 makes an async call back to Contract 1
+4. Shard 1 picks up the new transaction from Shard 2 and executes it
+5. Contract 1 performs whatever operation was requested by Contract 2
+
+As shown above, the async call inside Contract 2 effectively becomes a callback function for Contract 1. Note that the use of this pattern is only possible if Contract 1 or Contract 2 share the same owner (or their owners cooperate with each other).
+
+:::
+
+### Internal vs. external messages 
+
+In =nil;, messages are divided into internal and external.
+
+Internal messages cover all types of communication between smart contracts on the network regardless of the shard where they are deployed. When Contract 1 (deployed on Shard 1) calls a function in Contract 2 (deployed on Shard 2), Shard 2 spawns an internal message.
+
+
+External messages can originate only in one case: when any entity that exists outside of the cluster (e.g., a dApp) sends a message to any smart contract on the network.
+
+:::info[Payment for messages]
+
+For internal messages, the message itself is the 'payer' entity. An internal message spends its `value` to cover processing costs.
```

Comment:
It is only a payer for the top-level smart contract. If there's a chain of synchronous calls, nested calls will have payers other than the message itself.